internal/devices/assigned_devices.go

```diff
@@ -0,0 +1,162 @@
+// +build windows
+
+package devices
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"strings"
+
+	winio "github.com/Microsoft/go-winio"
+	"github.com/Microsoft/go-winio/pkg/guid"
+	"github.com/Microsoft/hcsshim/internal/cmd"
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/resources"
+	"github.com/Microsoft/hcsshim/internal/shimdiag"
+	"github.com/Microsoft/hcsshim/internal/uvm"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+)
+
+// handleAssignedDevicesWindows does all of the work to setup the hosting UVM and retrieve
+// device information for adding assigned devices on a WCOW container definition.
+//
+// First, devices are assigned into the hosting UVM. Drivers are then added into the UVM and
+// installed on the matching devices. This ordering allows us to guarantee that driver
+// installation on a device in the UVM is completed before we attempt to create a container.
+//
+// Then we find the location paths of the target devices in the UVM and return the results
+// as WindowsDevices.
+func HandleAssignedDevicesWindows(ctx context.Context, vm *uvm.UtilityVM, annotations map[string]string, specDevs []specs.WindowsDevice) (resultDevs []specs.WindowsDevice, closers []resources.ResourceCloser, err error) {
+	defer func() {
+		if err != nil {
+			// best effort clean up allocated resources on failure
+			for _, r := range closers {
+				if err := r.Release(ctx); err != nil {
+					log.G(ctx).WithError(err).Error("failed to release container resource")
+				}
+			}
+			closers = []resources.ResourceCloser{}
```

Comment:
Is this needed? closers would be gced after this function exits and we return nil in every error case regardless right? If this is needed for some reason shouldnt we just assign it to nil?

Reply:
We don't return nil in the case of an error from `getChildrenDeviceLocationPaths`. This was just to make it so that every error case does return nil for the closers.